### ISSUE TITLE
[clang-diagnostic-error] deduction of template arguments of 'concurrent_vector'

### ISSUE THREAD

**Issue by ianna** [cms-sw/cmssw#31664]
Using CMSSW_11_2_X_2020-10-03-1100 IB on slc7_amd64_gcc820 the following code compiles and runs fine:
```c++
    std::vector<double> results = splitNumeric(val);
    registry->insert({name, tbb::concurrent_vector(results.begin(), results.end()) });
```
however, clang fails with an error:
```bash
Error while processing /build/yana/DD4HEP/Performance/CMSSW_11_2_X_2020-10-03-1100/src/DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc.
/build/yana/DD4HEP/Performance/CMSSW_11_2_X_2020-10-03-1100/src/DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc:1677:29: error: no viable constructor or deduction guide for deduction of template arguments of 'concurrent_vector' [clang-diagnostic-error]
    registry->insert({name, tbb::concurrent_vector(results.begin(), results.end()) });
                            ^
```
[TBB Documentation](https://www.threadingbuildingblocks.org/docs/help/reference/containers_overview/concurrent_vector.html) states:
Where possible, constructors of concurrent_vector support class template argument deduction for C++17. For example
```c++
std::vector<int> v = {0, 1, 2};
tbb::concurrent_vector cv(v.begin(), v.end);
declares cv as tbb::concurrent_vector<int, tbb::cache_aligned_allocator<int>>.
```

**Comment by ianna** [comment]
@davidlange6 - FYI

**Comment by vgvassilev** [comment]
@ianna, can you provide a minimal example compiling with gcc and not with clang?

**Comment by ianna** [comment]
> @ianna, can you provide a minimal example compiling with gcc and not with clang?

```c++
std::vector<int> v = {0, 1, 2};
tbb::concurrent_vector cv(v.begin(), v.end);
```

**Comment by dan131riley** [comment]
The error for the constructor that should have matched is
```
/cvmfs/cms.cern.ch/slc7_amd64_gcc820/external/tbb/2019_U9/include/tbb/concurrent_vector.h:722:5: note: candidate template ignored: couldn't infer template argument 'T'
    concurrent_vector(I first, I last, const allocator_type &a = allocator_type())
```
Intel icc 2020.2 also fails to compile, using the Intel-supplied TBB, with the error for the relevant constructor
```
/opt/intel/compilers_and_libraries_2020.2.254/linux/tbb/include/tbb/concurrent_vector.h(722): note: this candidate was rejected because at least one template argument could not be deduced
      concurrent_vector(I first, I last, const allocator_type &a = allocator_type())
```

**Comment by dan131riley** [comment]
Type inference for that constructor needs a deduction guide,

https://github.com/oneapi-src/oneTBB/blob/eca91f16d7490a8abfdee652dadf457ec820cc37/include/tbb/concurrent_vector.h#L1169-L1175

```
// Deduction guide for the constructor from two iterators
template<typename I,
         typename T = typename std::iterator_traits<I>::value_type,
         typename A = cache_aligned_allocator<T>
> concurrent_vector(I, I, const A& = A())
-> concurrent_vector<T, A>;
```
and the define for `__TBB_CPP17_DEDUCTION_GUIDES_PRESENT` isn't getting set for clang.  This is supposed to be set here:

https://github.com/oneapi-src/oneTBB/blob/eca91f16d7490a8abfdee652dadf457ec820cc37/include/tbb/tbb_config.h#L237

but this Is evaluating to 0.  I think this is a TBB bug, clang defines `__cpp_deduction_guides` but doesn't consider it a feature--TBB should be testing for it similarly to the gcc test
```
#define __TBB_CPP17_DEDUCTION_GUIDES_PRESENT            (__cpp_deduction_guides >= 201606L)
```
I'll see about a bug report for TBB.

**Comment by makortel** [comment]
assign core

Thanks @dan131riley!

**Comment by dan131riley** [comment]
submitted https://github.com/oneapi-src/oneTBB/issues/284

**Comment by makortel** [comment]
From oneapi-src/oneTBB#284

> The issue with incorrect detection of Class Template Argument Deduction feature on Clang was fixed in oneTBB 2021.2.0.

**Comment by ianna** [comment]
@makortel and @dan131riley - please, open a new issue if it is still relevant. Thanks!
